### PR TITLE
Blazor native deps (w/callback coverage)

### LINKED ISSUE #23661: Document how to setup callbacks for native dependencies in Blazor WebAssembly
[EDIT by guardrex to add the metadata]

See https://github.com/dotnet/runtime/issues/60824#issuecomment-953035187.

@lewing @radical @lambdageek 

---
#### Document Details

⚠ *Do not edit this section. It is required for docs.microsoft.com ➟ GitHub issue linking.*

* ID: be59faf1-5a20-b7e9-19ff-06589023bb15
* Version Independent ID: 250e7458-3c1d-0117-261d-c03ac136f16a
* Content: [Host and deploy ASP.NET Core Blazor WebAssembly](https://docs.microsoft.com/en-us/aspnet/core/blazor/host-and-deploy/webassembly?view=aspnetcore-6.0#native-dependencies-support)
* Content Source: [aspnetcore/blazor/host-and-deploy/webassembly.md](https://github.com/dotnet/AspNetCore.Docs/blob/main/aspnetcore/blazor/host-and-deploy/webassembly.md)
* Product: **aspnet-core**
* Technology: **aspnetcore-hostdeploy**
* GitHub Login: @guardrex
* Microsoft Alias: **riande**

### PR DESCRIPTION
**Pull request by guardrex**
Fixes #23661

[Internal Review Topic](https://review.docs.microsoft.com/en-us/aspnet/core/blazor/webassembly-native-dependencies?view=aspnetcore-6.0&branch=pr-en-us-23668)

This guidance requires further work:

* The `WasmBuildNative` property wasn't in the blog post. I *guessed* where/how to describe it.
* If an example makes sense for the callback guidance, place it in an issue comment here for inclusion.
* I can't get the SkiaSharp example to work following this guidance with the GA SDK. AFAICT, this is local to my machine, but I'll need to test on another PC here to confirm. The example and guidance is straight out of the blog post AFAICT ...

  https://devblogs.microsoft.com/dotnet/asp-net-core-updates-in-net-6-rc-2/#using-libraries-with-native-dependencies

  The [SkiaSharp repo example](https://github.com/mono/SkiaSharp/tree/main/samples/Basic/Blazor/WebAssembly) isn't an exact match to the scenario because it doesn't use the NuGet package.

### CODE REVIEW

**Comment by mkArtakMSFT** [comment]
Thanks @guardrex.
@SteveSandersonMS can you please review this? Thanks!

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:22]: Could we be more explicit about its name here? Is it called ".NET WebAssembly build tools" in VS?

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:23]: Do you need to be admin? I thought I did it in the past without elevation.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:44]: You don't need this. As long as there's at least one `NativeFileReference`, it will do the right thing anyway.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (guardrex) [reply aspnetcore/blazor/webassembly-native-dependencies.md:22]: I'll check on the name. I've removed VS 2019 but haven't installed VS 2022 yet ... I'll do that today.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:51]: Also don't need to do this

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:62]: I'm surprised you don't need to mark the `fact` function as being exported in some way. As long as this code actually works as given then that's fine though.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:86]: In the final builds, debug relinking was sped up a lot and should now only take 5-10 seconds. So we could probably drop the statement that it "may take a few minutes".

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:89]: I guess this note can be removed now.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:102]: I think this section title is misleading. It says it's about using libraries, but the actual content is about *creating* libraries. People might mistakenly think that consuming a library like SkiaSharp involves doing stuff with architecture-specific directories, when they don't need to know or care about any of that.

Should the section title be changed to something like "Packaging native dependencies in a NuGet package"?

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (SteveSandersonMS) [inline aspnetcore/blazor/webassembly-native-dependencies.md:109]: > shouldn't be used in production.

That's a strong statement. I think what we mean to say is something like "do this if you want but don't blame us if there's a problem", whereas it sounds like "you should not do this because it's a bad thing to do". We're not actually trying to say there's something bad about the following code, right? We just mean to communicate that we're not officially promising anything about it.

Would it be enough simply to say "... is unsupported by Microsoft"? I think that communicates the intent.

**Review by SteveSandersonMS — APPROVED** [review]
I've made a bunch of suggestions but more broadly this looks great. Thanks @guardrex!

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (guardrex) [reply aspnetcore/blazor/webassembly-native-dependencies.md:23]: That's from @danroth27's blog post, but I agree. I did it here without an admin command shell.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (guardrex) [reply aspnetcore/blazor/webassembly-native-dependencies.md:62]: Again, it's from @danroth27 blog post. I've focused on the SkiaSharp example. I'll try this today before merging to make sure that it works.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (guardrex) [reply aspnetcore/blazor/webassembly-native-dependencies.md:109]: I yield to @danroth27. We have a few spots where we say that preview/experimental packages are not supported for production use. For example ...

> `Microsoft.AspNetCore.ProtectedBrowserStorage` is an unsupported, experimental package that isn't intended for production use.

> The `Microsoft.AspNetCore.Components.WebAssembly.MultipartBundle` NuGet package and the `CustomPackagedApp` sample app are unsupported, experimental demonstration resources not intended for production use.

... and IIRC, we did something similar for `AspNetCore.Diagnostics.HealthChecks` while it was in preview. My intention is to remove that part of the remark when SkiaSharp goes off of preview.

I'll 👂 @danroth27 for guidance on these examples and update them all on this PR.

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (guardrex) [reply aspnetcore/blazor/webassembly-native-dependencies.md:62]: Yes, this works as written. 🎉

File: aspnetcore/blazor/webassembly-native-dependencies.md
Comment (guardrex) [reply aspnetcore/blazor/webassembly-native-dependencies.md:106]: I'm asking Matthew if he has an ETA on a RTM/GA release at [Is there an ETA on RTM/GA for SkiaSharp.Views.Blazor? (mono/SkiaSharp #1862)](https://github.com/mono/SkiaSharp/issues/1862). I'll hold this PR for a sec to see if he says that the release is imminent. If so, I'll strike the warning about production use and leave our usual 'not supported by MS' remark.

Comment (danroth27) [reply aspnetcore/blazor/webassembly-native-dependencies.md:106]: Users are free to use whatever features/code they want in production, as long as they've tested that they satisfy their needs. Pretty much all of .NET, stable or not, is provided "as is" as stated in our various licenses. So, we really only need to clarify the level of support and intent to release a stable version.

Suggested language: "Experimental and preview features are provided for the purpose of collecting feedback and are not supported for production use." 

The difference between experimental and preview features is that preview features are expect to eventual ship a stable release, while experimental features may not. Suggested language: "Experimental features are provided for the purpose of exploring feature viability and may not ship a future stable version".

**Pull request merged**
